### PR TITLE
Remove dependency from OpenTelemetry.Instrumentation.SqlClient

### PR DESCRIPTION
**Pull request by sebastienros**
Used two a separate forlder for `/Shared` since I expect the other OTel libraries we need to include will need this too.

See https://github.com/dotnet/aspire/issues/2819
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3059)

### CODE REVIEW

File: src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/GlobalSuppressions.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// This file is used by Code Analysis to maintain SuppressMessage
+// attributes that are applied to this project.
+// Project-level suppressions either have no target or are given
+// a specific target and scoped to a namespace, type, member, etc.
+
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("Style", "IDE0028:Simplify collection initialization", Justification = "Vendor code", Scope = "namespaceanddescendants", Target = "~N:OpenTelemetry.Instrumentation.SqlClient")]
```
Comment (eerhardt) [inline src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/GlobalSuppressions.cs:8]: Can we just use .editorconfig for this?

File: THIRD-PARTY-NOTICES.TXT [hunk 2]
```diff
@@ -57,3 +57,20 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
+
+License notice for OpenTelemetry .NET (https://github.com/open-telemetry/opentelemetry-dotnet)
```
Comment (eerhardt) [inline THIRD-PARTY-NOTICES.TXT:61]: note that the next one (StackExchange.Redis) comes from a different repo: https://github.com/open-telemetry/opentelemetry-dotnet-contrib. We may need to update this when we pull that code in.

File: src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/SqlClientTraceInstrumentationOptions.cs [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+
+namespace OpenTelemetry.Instrumentation.SqlClient;
+
+/// <summary>
+/// Options for <see cref="SqlClientInstrumentation"/>.
+/// </summary>
+/// <remarks>
+/// For help and examples see: <a href="https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Instrumentation.SqlClient/README.md#advanced-configuration" />.
+/// </remarks>
+public class SqlClientTraceInstrumentationOptions
```
Comment (eerhardt) [inline src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/SqlClientTraceInstrumentationOptions.cs:19]: These need to be `internal`. We don't want to ship these as public from our library.

**Comment by eerhardt** [comment]
@reyang - Can you review the attribution to ensure we aren't violating the license here?

File: THIRD-PARTY-NOTICES.TXT (hunk 2 above)
Comment (reyang) [reply THIRD-PARTY-NOTICES.TXT:61]:
```suggestion
License notice for OpenTelemetry
```

File: THIRD-PARTY-NOTICES.TXT [hunk 4]
```diff
@@ -57,3 +57,20 @@ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
...
+----------------------------------------------------------------------------------------------
+
+Copyright 2023 OpenTelemetry Authors
```
Comment (reyang) [inline THIRD-PARTY-NOTICES.TXT:64]: This is what OpenTelemetry uses https://github.com/open-telemetry/opentelemetry-dotnet/blob/af57ae87871aac7e201a03255ac8c81e5747bc07/src/OpenTelemetry.Api/Baggage.cs#L1-L2

```suggestion
Copyright The OpenTelemetry Authors
```

File: src/Vendoring/README.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# Vendoring code sync instructions
```
Comment (reyang) [inline src/Vendoring/README.md:1]: @rajkumar-rangaraj FYI

**Comment by reyang** [comment]
> @reyang - Can you review the attribution to ensure we aren't violating the license here?

I've added couple suggestions, it looks good to me once these are addressed 👍

**Review by joperezr — COMMENTED** [review]
So we decided to go with building their code ourselves as opposed to re-disting their libraries inside our package?

This makes me nervous as we won't be running their tests in this code, nor would we be applying code fixes, and we don't really know how long are we going to be doing this.

Did we explore the option of just harvesting their library and including it in our package instead?

cc @eerhardt

File: src/Vendoring/README.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+### Instructions
+
+- Copy files from `src/OpenTelemetry.Instrumentation.SqlClient`:
+    - `OpenTelemetry.Instrumentation.SqlClient.csproj`
+    - `**\*.cs`
+
+### Customizations
+
+- Added `#nullable disable` in files that require it.
+- Added `GlobalSuppressions.cs` to fix incompatible coding style.
```
Comment (joperezr) [inline src/Vendoring/README.md:32]: We should also add here that all public types need to be made internal. As @eerhardt pointed out we need to make sure we are not exposing any of this as public API.

File: src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+using OpenTelemetry.Instrumentation.SqlClient;
+using OpenTelemetry.Instrumentation.SqlClient.Implementation;
+using OpenTelemetry.Internal;
+
+namespace OpenTelemetry.Trace;
+
+/// <summary>
+/// Extension methods to simplify registering of dependency instrumentation.
+/// </summary>
+public static class TracerProviderBuilderExtensions
```
Comment (joperezr) [inline src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/TracerProviderBuilderExtensions.cs:19]: Needs to be internal

File: src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/SqlClientTraceInstrumentationOptions.cs (hunk 3 above)
Comment (reyang) [reply src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/SqlClientTraceInstrumentationOptions.cs:19]: One suggestion for future releases, consider leveraging the public API analyzer in CI.

* https://github.com/open-telemetry/opentelemetry-dotnet/pull/806
* https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Api/.publicApi/Stable/PublicAPI.Shipped.txt

File: src/Vendoring/README.md (hunk 6 above)
Comment (reyang) [reply src/Vendoring/README.md:32]: https://github.com/dotnet/aspire/pull/3059#discussion_r1534387563

File: src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/GlobalSuppressions.cs (hunk 1 above)
Comment (sebastienros) [reply src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/GlobalSuppressions.cs:8]: These are specific to the files from OTel, but I will check if we can set rules for specific paths.

File: THIRD-PARTY-NOTICES.TXT (hunk 4 above)
Comment (sebastienros) [reply THIRD-PARTY-NOTICES.TXT]: Is it worth filing an issue in Azure SDK? I took it from there: https://github.com/Azure/azure-sdk-for-net/blob/main/NOTICE.txt#L232-L235

Comment (reyang) [reply THIRD-PARTY-NOTICES.TXT]: > Is it worth filing an issue in Azure SDK? I took it from there: https://github.com/Azure/azure-sdk-for-net/blob/main/NOTICE.txt#L232-L235

@rajkumar-rangaraj could you follow up and handle this?

Comment (rajkumar-rangaraj) [reply THIRD-PARTY-NOTICES.TXT]: I will get this handled in Azure SDK. 
When we earlier vendored in Azure SDK, OpenTelemetry had that license format.

It's worth to update the Readme to check on the license formats too.

File: src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/GlobalSuppressions.cs (hunk 1 above)
Comment (sebastienros) [reply src/Vendoring/OpenTelemetry.Instrumentation.SqlClient/GlobalSuppressions.cs]: Done

File: src/Vendoring/README.md (hunk 6 above)
Comment (sebastienros) [reply src/Vendoring/README.md:32]: Done

**Comment by eerhardt** [comment]
> So we decided to go with building their code ourselves as opposed to re-disting their libraries inside our package?
> 
> This makes me nervous as we won't be running their tests in this code, nor would we be applying code fixes, and we don't really know how long are we going to be doing this.
> 
> Did we explore the option of just harvesting their library and including it in our package instead?
> 
> cc @eerhardt

Redisting libraries in a nuget package isn't without drawbacks either. If the app brings in the OpenTelemetry.Instrumentation.SqlClient package through another dependency, there will be conflicts since 2 nuget packages brought in the same assembly. We would also need to sign it, etc, in order to redist it.

> and we don't really know how long are we going to be doing this.

My understanding from @reyang is that the SqlClient library is targeted to ship stable later this calendar year. But I assume those plans can change. The other one we will need to give the same treatment is the StackExchange.Redis OTel library, which doesn't have a plan on going stable AFAIK.

File: src/Components/Aspire.Microsoft.Data.SqlClient/Aspire.Microsoft.Data.SqlClient.csproj [hunk 8]
```diff
@@ -15,15 +15,19 @@
...
+    <Compile Include="$(VendoringDir)OpenTelemetry.Shared\**\*.cs" LinkBase="OpenTelemetry.Instrumentation.SqlClient.Shared" />
+  </ItemGroup>
+
   <ItemGroup>
     <PackageReference Include="AspNetCore.HealthChecks.SqlServer" />
-    <PackageReference Include="Microsoft.Data.SqlClient" />
     <PackageReference Include="Microsoft.Extensions.Configuration.Binder" />
     <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks" />
     <PackageReference Include="Microsoft.Extensions.Hosting.Abstractions" />
+    <PackageReference Include="OpenTelemetry.Instrumentation.AspNetCore" />
```
Comment (eerhardt) [inline src/Components/Aspire.Microsoft.Data.SqlClient/Aspire.Microsoft.Data.SqlClient.csproj:28]: What's this?

File: src/Components/Aspire.Microsoft.Data.SqlClient/Aspire.Microsoft.Data.SqlClient.csproj [hunk 9]
```diff
@@ -15,15 +15,19 @@
...
   </ItemGroup>
 
+  <ItemGroup>
+    <Compile Include="$(VendoringDir)OpenTelemetry.Instrumentation.SqlClient\**\*.cs" LinkBase="OpenTelemetry.Instrumentation.SqlClient" />
+    <Compile Include="$(VendoringDir)OpenTelemetry.Shared\**\*.cs" LinkBase="OpenTelemetry.Instrumentation.SqlClient.Shared" />
+  </ItemGroup>
+
   <ItemGroup>
     <PackageReference Include="AspNetCore.HealthChecks.SqlServer" />
-    <PackageReference Include="Microsoft.Data.SqlClient" />
```
Comment (eerhardt) [inline src/Components/Aspire.Microsoft.Data.SqlClient/Aspire.Microsoft.Data.SqlClient.csproj:20]: We want this, so our dependency on Microsoft.Data.SqlClient gets our version from the Directory.Packages.props.

Comment (sebastienros) [reply src/Components/Aspire.Microsoft.Data.SqlClient/Aspire.Microsoft.Data.SqlClient.csproj:20]: Was not even on purpose, I don't know how I removed it. Maybe mistaken for the telemetry one.

File: src/Components/Aspire.Microsoft.Data.SqlClient/Aspire.Microsoft.Data.SqlClient.csproj (hunk 8 above)
Comment (sebastienros) [reply src/Components/Aspire.Microsoft.Data.SqlClient/Aspire.Microsoft.Data.SqlClient.csproj]: I think when I imported some source this was required for a specific type. Now not anymore so I assume that I removed a file that was the culprit. Removed.

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks!

**Review by reyang — APPROVED** [review]

**Pull request merged**